webhooks/virtualmachinegrouppublishrequest/validation/virtualmachinegrouppublishrequest_validator.go

```diff
@@ -0,0 +1,161 @@
+// validateCreateSpec ensures spec.source, spec.target, and spec.virtualMachines are set.
+// Those are optional fields but will be set by the mutating webhook if they are omitted.
+func (v validator) validateCreateSpec(ctx *pkgctx.WebhookRequestContext,
+	groupPubReq *vmopv1.VirtualMachineGroupPublishRequest) field.ErrorList {
+	var fieldErrs field.ErrorList
+	specPath := field.NewPath("spec")
+
+	vmGroupSource := &vmopv1.VirtualMachineGroup{ObjectMeta: metav1.ObjectMeta{Namespace: groupPubReq.Namespace,
+		Name: groupPubReq.Spec.Source}}
+	clTarget := &imgregv1a1.ContentLibrary{ObjectMeta: metav1.ObjectMeta{Namespace: groupPubReq.Namespace,
+		Name: groupPubReq.Spec.Target}}
+
+	fieldErrs = append(fieldErrs, validateResource(ctx, v.client, specPath.Child("source"), vmGroupSource))
+	fieldErrs = append(fieldErrs, validateResource(ctx, v.client, specPath.Child("target"), clTarget))
+	fieldErrs = append(fieldErrs, v.validateSpecVMs(specPath, groupPubReq))
+
+	return fieldErrs
+}
+
+func (v validator) ValidateDelete(_ *pkgctx.WebhookRequestContext) admission.Response {
+	return admission.Allowed("")
+}
+
+func (v validator) ValidateUpdate(ctx *pkgctx.WebhookRequestContext) admission.Response {
+	groupPubReq, err := v.vmGroupPublishRequestFromUnstructured(ctx.Obj)
+	if err != nil {
+		return webhook.Errored(http.StatusBadRequest, err)
+	}
+	oldGroupPubReq, err := v.vmGroupPublishRequestFromUnstructured(ctx.OldObj)
+	if err != nil {
+		return webhook.Errored(http.StatusBadRequest, err)
+	}
+
+	return common.BuildValidationResponse(ctx, nil,
+		common.ConvertFieldErrorsToStrings(v.validateUpdateSpec(groupPubReq, oldGroupPubReq)), nil)
+}
+
+// validateUpdateSpec ensures no updates to spce immutable fields.
+func (v validator) validateUpdateSpec(groupPubReq, oldGroupPubReq *vmopv1.VirtualMachineGroupPublishRequest) field.ErrorList {
+	var fieldErrs field.ErrorList
+	specPath := field.NewPath("spec")
+
+	fieldErrs = append(fieldErrs, validation.ValidateImmutableField(groupPubReq.Spec.Source, oldGroupPubReq.Spec.Source, specPath.Child("source"))...)
+	fieldErrs = append(fieldErrs, validation.ValidateImmutableField(groupPubReq.Spec.Target, oldGroupPubReq.Spec.Target, specPath.Child("target"))...)
+	fieldErrs = append(fieldErrs, validation.ValidateImmutableField(groupPubReq.Spec.VirtualMachines, oldGroupPubReq.Spec.VirtualMachines, specPath.Child("virtualMachines"))...)
+
+	return fieldErrs
+}
+
+// validateSpecVMs ensures spec.VirtualMachines is defined either via the user or the mutation webhook.
+// It does not check the existence of those VMs since it would require multiple calls to the API server.
+// Rather, each VM publish request's status is reflected on VM group publish request's status.
+func (validator) validateSpecVMs(specPath *field.Path, groupPubReq *vmopv1.VirtualMachineGroupPublishRequest) *field.Error {
+	if len(groupPubReq.Spec.VirtualMachines) == 0 {
+		return field.Required(specPath.Child("virtualMachines"), mutatingWebhookDownWarning)
```

Comment:
Maybe the mutation webhook should fail instead with a more specific error message.

And then this validation webhook fails if this is empty.